Pup 6841 document parser api

hlindberg opened the pull request:
PR contains a draft of the documentation (to end up somewhere else) and an example that makes use of the API.
PR made to make it easier to review and comment.

hlindberg committed:
(PUP-6841) Add documentation and example of Puppet Parser API

thallgren commented on parser_api/example.rb:
```diff
@@ -0,0 +1,173 @@
+  class MyChecker
+    attr_reader :acceptor
+    def initialize(diagnostics_producer)
+      @@bad_word_visitor       ||= Puppet::Pops::Visitor.new(nil, "badword", 0, 0)
```
The concept of the visitor/polymorphic dispatcher must be explained. To go from this line what actually happens (calls to `badword_QualifiedName` etc.) is not self evident.

hlindberg replied:
Ok, will explain.

hlindberg replied:
Added some additional text, but mainly referred to the blog post I wrote 2014.

thallgren commented on parser_api/example.rb:
```diff
@@ -0,0 +1,173 @@
+      check_bad_word(model)
+
+      # Then check all of its content
+      model.eAllContents.each {|m| check_bad_word(m) }
```
This exposes Rgen. It might be of value to refrain from doing that in examples if we want to get rid of Rgen later on.

hlindberg replied:
Don't think we have an alternative to eAllContents that is available from 3.8 future parser and forward?
And I think we should say that only eAllContent is API to reduce the amount of exposure to RGen API that we may not want to keep in an optimized implementation,

thallgren commented on parser_api/example.rb:
```diff
@@ -0,0 +1,173 @@
+result = result.model
+
+# validate using the default validator and get hold of the acceptor containing the result
+acceptor = parser.validate(result)
```
Perhaps make the `validate` method capable of validating the _result_ from `parse_string` rather than the model that it contains. That way, the _result_ becomes opaque (i.e. no need to disclose the somewhat unintuitive implementation detail `result = result.model` ).

hlindberg replied:
Sure, but the fact that it returns a Factory for further manipulation remains and a user of the API needs to know as they are bound to stumble over that otherwise,
(Potentially a lot of stuff (tests) to rewrite I think if we do not return a Factory, so don't think that is a good option either even if it is simple to add a Factory wrapper if one is needed).

thallgren commented on parser_api/example.rb:
```diff
@@ -0,0 +1,173 @@
+#
+puts "Standard validation errors found: #{acceptor.error_count}"
+puts "Standard validation warnings found: #{acceptor.warning_count}"
+
```
I would start with this example since it shows how to do a complete parse/validate using what's "in the box". Once that's covered, it's easier to understand that the objective with the `MyValidation` module is to go beyond that. It's also easier to understand how things fit together.

hlindberg replied:
Ok, will do a better "set up" before presenting all of the example code.

hlindberg replied:
I did the setup in the document rather than in the runnable code

hlindberg committed:
(PUP-6841) Add more info about polymorphic dispatch

Added link to blog post about polymorphic dispatch.

hlindberg committed:
(PUP-6841) Add link to more info about location processing

hlindberg committed:
(PUP-6841) Add link to AdapterPattern blog post.

hlindberg committed:
(PUP-6841) Switch example and the impl it uses around for readability

This puts the example use before the example implementation to make it
easier to understand the purpose of the implementation.

hlindberg commented on parser_api/parser_api.md:
```diff
@@ -0,0 +1,1064 @@
+
+The result of parsing is an instance of `Puppet::Pops::Model::Factory`, on which there is one public method named `model` that returns the Abstract Syntax Tree (AST) that represents the parsed source. The root of this tree is always an instance of
+`Puppet::Pops::Model::Program`. This class, along with the other close to 90 classes that make up all kinds of expressions and statements in the puppet language are found in the file `lib/puppet/pops/model/model_meta.rb`
+
```
The above changes in Puppet 5.0.0 so that there is no need to unwrap the returned factory using the call to `model`. We added a backwards compatible method which still allows calling `model` on the return value, that then just returns self.

hlindberg commented:
This PR needs an update for Puppet 5.0.0 since some of the methods (like `eAllContents`) have changed.

jeremy-asher commented:
Thanks for writing this up @hlindberg!  The detailed example is a huge help.  I'd love to hear how this could be adapted for Puppet 5 when you have a chance.

dinhthang5 commented:
Pls help
